v2 segment support string encode(#1766)

wangbo opened the pull request:
#1766
major change
1. change data format of binary dict page, appending (dict page data) and (dict page offset) to binary dict page; 
2. add new decoding method for new binary dict page format
3. add ut for segment test
5. set the elements of initial array to 0 ,when calling arena.AllocateNewBlock
6. hard code way to choose dict coding for string

0919 commit major change
1. change dict file format:when saving binary dict page, separate dict page from dict page,one dict page may have multi data pages;when reading a binary dict page,one ColumnReader keeps one dict page
2. loading dict when calling column_reader._read_page
3.rollback BinaryDictPage
4. no longer using memset(0) to inital column_zonemap.max_value

0926 17 commit major change
1. init column_zone_map min value column_zone_map slice's data array;
    set char/varchar column_zone_map'max value size to 0
2. add ut for char column zone map query hit/miss

0929 10 commit major change
1. allocate mem for column_zone_map 's max and min value
2. direct copy content to column_zone_map's max and min value

wangbo committed:
v2 segment support string encode(#1766)

wangbo committed:
Merge remote-tracking branch 'upstream/master'

wangbo committed:
segment test using uint64

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -102,7 +102,16 @@ Slice BinaryDictPageBuilder::finish() {
+    if (_encoding_type == DICT_ENCODING) {
+        size_t dict_offset = _buffer.size();
+        Slice dictionary_page;
+        get_dictionary_page(&dictionary_page);
```
there may be memory leak. You should delete dictionary_page.data after usage

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -173,6 +189,7 @@ Status BinaryDictPageDecoder::init() {
     }
 
     RETURN_IF_ERROR(_data_page_decoder->init());
+    RETURN_IF_ERROR(_dict_decoder->init());
```
if encoding_type is plain, _dict_decoder is nullptr. That will lead to crash

wangbo replied:
👌

imay commented on be/src/util/arena.cpp:
```diff
@@ -59,6 +60,7 @@ char* Arena::AllocateAligned(size_t bytes) {
 
 char* Arena::AllocateNewBlock(size_t block_bytes) {
     char* result = new char[block_bytes];
+    memset(result, 0, block_bytes);
```
Why memset this to 0? This will cost CPU, we should avoid unnecessary memset

wangbo replied:
initialing column_zone_map.max_value will call this method for allocating memory,and then column_zone_map.max_value will be reinterpret_cast to slice(FieldTypeTraits<OLAP_FIELD_TYPE_CHAR>.set_to_min,types.h),if not set to 0,the slice.size value will be unknown(usually very large value), it will cause core dump;
maybe I  cast column_zone_map.max_value to slice and set to 0 instead of  memset to 0 is better

imay replied:
Yes, I think it's better to only fix column_zone_map's problem, because this is used in many other code in which memset is unnecessary.

wangbo replied:
👌

wangbo committed:
update commit

wangbo committed:
update zone map

imay commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -166,6 +167,24 @@ void ColumnReader::get_row_ranges_by_zone_map(CondColumn* cond_column, RowRanges
     _calculate_row_ranges(page_indexes, row_ranges);
 }
 
+Status ColumnReader::get_dict_page_decoder(BinaryDictPageDecoder* binaryDictPageDecoder) {
```
should return _column_dict_page_decoder, not set _column_dict_page_decoder in this function

wangbo replied:
return status here can make FileColumnIterator returns error when read_page failed or dict_page_decoder-> init failed;

gaodayue replied:
ColumnReader is shared by multiple FileColumnIterator and thus should be thread-safe.

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/column_reader.h:
```diff
@@ -103,6 +106,9 @@ class ColumnReader {
     std::unique_ptr<ColumnZoneMap> _column_zone_map;
+
+    // keep dict page
+    std::shared_ptr<BinaryPlainPageDecoder> _column_dict_page_decoder;
```
why shared_ptr? I think unique_ptr is enough.

wangbo replied:
one dict page may correspond to many data pages,so _column_dict_page_decoder ptr may be share by many BinaryDictPageDecoder

gaodayue replied:
ColumnReader is owned by Segment and Rowset, being alive even when there is no query. Therefore it should retain memory as small as possible. I think it'd better to keep dict page decoder in `FileColumnIterator`.

wangbo replied:
👌

gaodayue commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -166,6 +167,24 @@ void ColumnReader::get_row_ranges_by_zone_map(CondColumn* cond_column, RowRanges
+        Slice dict_data = ph.data();
+
+        std::shared_ptr<BinaryPlainPageDecoder> dict_page_decoder(
+                new BinaryPlainPageDecoder(dict_data));
```
To retain memory for dict page, you should also hold `PageHandle`, otherwise PageCache could evict the page referenced by dict_page_decoder.

wangbo replied:
👌

gaodayue commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -411,6 +430,13 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
     RETURN_IF_ERROR(page->data_decoder->init());
 
+    if (_reader->encoding_info()->encoding() == DICT_ENCODING) {
+        BinaryDictPageDecoder* binary_dict_page_decoder = (BinaryDictPageDecoder*)page->data_decoder;
```
As mentioned above, BinaryDictPageDecoder should got dict page via constructor.

gaodayue commented on be/src/olap/rowset/segment_v2/segment_writer.cpp:
```diff
@@ -60,6 +60,9 @@ Status SegmentWriter::init(uint32_t write_mbytes_per_sec) {
 
         ColumnWriterOptions opts;
+        if (column.type() == OLAP_FIELD_TYPE_CHAR || column.type() == OLAP_FIELD_TYPE_VARCHAR) {
+            opts.encoding_type = DICT_ENCODING;
```
You don't need to explicitly set encoding_type here because the default encoding for char and varchar is DICT_ENCODING.

wangbo replied:
👌

gaodayue replied:
forgot to remove?

wangbo replied:
content rollback when merging branch

wangbo commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -116,6 +116,10 @@ class BinaryDictPageDecoder : public PageDecoder {
 
+    bool is_dict_encoding();
+
+    void set_dict_decoder(std::shared_ptr<BinaryPlainPageDecoder> dict_decoder);
```
using BinaryDictPageDecoder's constructor to set dict page decoder means first need to decode slice data header to check whether current page is dict encoding or not,so there is two ways to do it:
1. decode slice data header in FileColumnIterator and getting the encode type,construct dict page decoder and set it to PageDecoderOptions;It works but "decode slice data header" will be done twice
2. decode slice data header and settting the encode type PageDecoderOptions, BinaryDictPageDecoder's no longer to decode the slice data header;but this will change the logic of BinaryDictPageDecoder->init

so I think first call BinaryDictPageDecoder->init to get encoding type and then set dict page decoder to BinaryDictPageDecoder is better

gaodayue replied:
I see. Passing dict page decoder via constructor will force us to always read dict page even when only non-dict-encoded data pages are seeked.

I agree with you now to lazy set dict decoder via member function. Two more suggestions

1. remove `dict_decoder` from PageDecoderOptions
2. the argument type should be `PageDecoder*`, since dict decoder is owned by FileColumnIterator

wangbo replied:
👌

wangbo committed:
update commit 0924:
keep dict page handle in FileColumnIterator;

wangbo committed:
Merge remote-tracking branch 'upstream/master' into master_test

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -116,11 +116,15 @@ class BinaryDictPageDecoder : public PageDecoder {
     PageDecoderOptions _options;
     std::unique_ptr<PageDecoder> _data_page_decoder;
-    std::shared_ptr<BinaryPlainPageDecoder> _dict_decoder;
+    BinaryPlainPageDecoder* _dict_decoder;
```
```suggestion
    const BinaryPlainPageDecoder* _dict_decoder;
```

wangbo replied:
👌

gaodayue commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -426,6 +431,23 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
+                RETURN_IF_ERROR(dict_decoder->init());
+
+                std::unique_ptr<PageDecoder> dict_page_ptr(dict_decoder);
+                _dict_decoder = std::move(dict_page_ptr);
```
prefer `_dict_decoder.reset(new BinaryPlainPageDecoder(_dict_page_handle.data()))`

gaodayue commented on be/src/olap/rowset/segment_v2/column_reader.h:
```diff
@@ -29,6 +29,7 @@
 #include "olap/rowset/segment_v2/ordinal_page_index.h" // for OrdinalPageIndexIterator
 #include "olap/rowset/segment_v2/column_zone_map.h" // for ColumnZoneMap
 #include "olap/rowset/segment_v2/row_ranges.h" // for RowRanges
+#include "page_handle.h"
```
use full import path

wangbo replied:
👌

gaodayue commented on be/test/olap/rowset/segment_v2/binary_dict_page_test.cpp:
```diff
@@ -53,17 +53,17 @@ class BinaryDictPageTest : public testing::Test {
         PageDecoderOptions dict_decoder_options;
-        std::shared_ptr<BinaryPlainPageDecoder> dict_page_decoder(
-                new BinaryPlainPageDecoder(dict_slice, dict_decoder_options));
+        BinaryPlainPageDecoder* dict_page_decoder = new BinaryPlainPageDecoder(dict_slice, dict_decoder_options);
```
use unique_ptr to avoid memory leak

wangbo replied:
👌

gaodayue commented on be/test/olap/rowset/segment_v2/binary_dict_page_test.cpp:
```diff
@@ -147,16 +147,15 @@ class BinaryDictPageTest : public testing::Test {
             PageDecoderOptions dict_decoder_options;
-            std::shared_ptr<BinaryPlainPageDecoder> dict_page_decoder(
-                    new BinaryPlainPageDecoder(dict_slice, dict_decoder_options));
+            BinaryPlainPageDecoder* dict_page_decoder = new BinaryPlainPageDecoder(dict_slice, dict_decoder_options);
```
use unique_ptr to avoid memory leak

wangbo commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -426,6 +431,23 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
+                RETURN_IF_ERROR(_reader->read_page(pp, &_dict_page_handle));
+
+                BinaryPlainPageDecoder* dict_decoder = new BinaryPlainPageDecoder(_dict_page_handle.data());
+                RETURN_IF_ERROR(dict_decoder->init());
```
👌

wangbo committed:
update commit 0925 10

gaodayue commented on be/src/olap/rowset/segment_v2/column_reader.h:
```diff
@@ -85,6 +86,8 @@ class ColumnReader {
     void get_row_ranges_by_zone_map(CondColumn* cond_column,
             const std::vector<CondColumn*>& delete_conditions, RowRanges* row_ranges);
 
+    PagePointer get_dict_page_pointer();
```
```suggestion
    PagePointer get_dict_page_pointer() const;
```

wangbo replied:
👌

gaodayue commented on be/test/olap/rowset/segment_v2/segment_test.cpp:
```diff
@@ -606,6 +606,173 @@ TEST_F(SegmentReaderWriterTest, TestDefaultValueColumn) {
+
+        auto* dest_slice = (Slice*)target;
+        dest_slice->size = _length;
+        dest_slice->data = new char[dest_slice->size];
```
You could use an `Arena` to allocate string data, otherwise you got memory leak

wangbo replied:
👌

gaodayue commented on be/test/olap/rowset/segment_v2/segment_test.cpp:
```diff
@@ -606,6 +606,173 @@ TEST_F(SegmentReaderWriterTest, TestDefaultValueColumn) {
+        memcpy(dest_slice->data, src_value, src_len);
+        memset(dest_slice->data + src_len, 0, dest_slice->size - src_len);
+    } else if (fieldType == OLAP_FIELD_TYPE_VARCHAR) {
+        Slice* slice = new Slice(*new string(&std::to_string(src)[0]));
```
same as above, use `Arena` to hold string data. In addition, never allocate Slice or string on heap.

wangbo replied:
In a for loop,if aollocate string on stack,and using memcpy to copy data to a char array,then all the elements in char array will be same and the value is the last value;
I think this because all the

gaodayue commented on be/test/olap/rowset/segment_v2/segment_test.cpp:
```diff
@@ -606,6 +606,173 @@ TEST_F(SegmentReaderWriterTest, TestDefaultValueColumn) {
+    // 10, 11, 12, 13
+    // 20, 21, 22, 23
+    //
+    // 64k int will generate 4 pages
```
delete the misleading comment

wangbo replied:
👌

gaodayue commented on be/test/olap/rowset/segment_v2/segment_test.cpp:
```diff
@@ -606,6 +606,173 @@ TEST_F(SegmentReaderWriterTest, TestDefaultValueColumn) {
+        for (int j = 0; j < 4; ++j) {
+            auto cell = row.cell(j);
+            cell.set_not_null();
+            set_column_value_by_type(tablet_schema->_cols[j]._type, i * 10 + j, (char*)cell.mutable_cell_ptr(), tablet_schema->_cols[j]._length);
```
I'm confused, how can you access private field `_cols` and `_type`?

wangbo replied:
me too;not FRIEND_TEST,not "#define private public"

gaodayue commented on be/test/olap/rowset/segment_v2/segment_test.cpp:
```diff
@@ -606,6 +606,173 @@ TEST_F(SegmentReaderWriterTest, TestDefaultValueColumn) {
+                        ASSERT_FALSE(BitmapTest(column_block.null_bitmap(), i));
+                        const Slice* actual = reinterpret_cast<const Slice*>(column_block.cell_ptr(i));
+
+                        char* expect = new char[sizeof(Slice)];
```
```suggestion
                        Slice expect;
```

wangbo replied:
👌

wangbo committed:
update commit 0925 8

wangbo committed:
Merge remote-tracking branch 'upstream/master'

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -116,11 +116,15 @@ class BinaryDictPageDecoder : public PageDecoder {
     PageDecoderOptions _options;
     std::unique_ptr<PageDecoder> _data_page_decoder;
-    std::shared_ptr<BinaryPlainPageDecoder> _dict_decoder;
+    const BinaryPlainPageDecoder* _dict_decoder;
```
initialize this member

wangbo replied:
mean set to nullptr?

kangpinghuang replied:
yes

kangpinghuang commented on be/src/olap/rowset/segment_v2/column_zone_map.cpp:
```diff
@@ -82,6 +82,8 @@ void ColumnZoneMapBuilder::_reset_zone_map() {
     min_slice->data = _max_string_value;
     min_slice->size = OLAP_STRING_MAX_LENGTH;
+    Slice *max_slice = (Slice *)_zone_map.max_value;
+    max_slice->size = 0;
```
do not do this, because the column maybe not char/varchar type.
and set_to_min(_zone_map.max_value) will set the initial value of max_value to min.

wangbo replied:
max_value/min_value 's memory comes char array of arena,but the content of char array is unknown,it's necessary to init the array first,or the size of slice will be very large, then core dump happends

wangbo replied:
1. the char array allocated to min_value and max_value must be initialed(like set to 0),otherwise when casting array to slice ,the size property of slice may be very huge
2. I think do the init in column_zone_map is better,because column_zone_map contains the context(like _max_string_value,arena),only do init when type=char

kangpinghuang commented on be/src/olap/rowset/segment_v2/encoding_info.cpp:
```diff
@@ -67,6 +68,18 @@ struct TypeEncodingTraits<type, RLE> {
+struct TypeEncodingTraits<type, DICT_ENCODING> {
+    static Status create_page_builder(const PageBuilderOptions& opts, PageBuilder** builder) {
+        *builder = new BinaryDictPageBuilder(opts);
+        return Status::OK();
```
maybe this type function can return void

wangbo replied:
agree ,it seems that there is no RETURN_IF_ERROR logic in class constructor;
And I think it's  better to make a new commit to fix it.

wangbo committed:
update commit 0926:17

kangpinghuang commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -426,6 +431,20 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
+    if (_reader->encoding_info()->encoding() == DICT_ENCODING) {
+        BinaryDictPageDecoder* binary_dict_page_decoder = (BinaryDictPageDecoder*)page->data_decoder;
+        if (binary_dict_page_decoder->is_dict_encoding()) {
+            if (_dict_decoder == nullptr) {
```
I think move this new dict_decoder  logic to column_reader because all column iterator use the same dict_decoder and the column reader has to load the dict if the column is dict encoding.

wangbo replied:
for high cardinality column,dict encoding will fallback to plain encoding,its data will contain both encoding, so I think loading dict after data page init and knows the real encoding is better

gaodayue replied:
Hi kangping, the reasons for reading dict page inside iterator are

1. Even if a column use dictionary encoding, there could be cases when only non-dict-encoded data pages are seeked. So we lazy read dict page only when necessary
2. `ColumnReader` is owned by Segment and Rowset, being alive even when there is no query. It should retain memory as small as possible.
3. Iterators of the same column won't repeat load the dict page because of page cache.

kangpinghuang replied:
ok, memory is a problems

kangpinghuang commented on gensrc/proto/segment_v2.proto:
```diff
@@ -95,7 +95,7 @@ message ColumnMetaPB {
 
     // // dictionary page for DICT_ENCODING
-    // optional PagePointerPB dict_page = 2;
+    optional PagePointerPB dict_page = 9;
```
where to call set_dict_page?
Have you missed it in segment_writer?

wangbo committed:
re fresh commit

wangbo committed:
reuse column zone map char array

wangbo committed:
reuse column zone map char array

wangbo committed:
add get_type_value_with_arena

wangbo committed:
Merge branch 'upstream_master' into master_test

kangpinghuang commented on be/src/olap/rowset/segment_v2/column_zone_map.cpp:
```diff
@@ -79,10 +87,17 @@ Status ColumnZoneMapBuilder::flush() {
+    } else if (_type_info->type() == OLAP_FIELD_TYPE_CHAR) {
+        Slice *min_value = (Slice *)_zone_map.min_value;
+        min_value->data = _max_char_value;
+        min_value->size = OLAP_CHAR_MAX_LENGTH;
```
why set char type's max length to 255? I think it should be equal to OLAP_STRING_MAX_LENGTH.

wangbo replied:
Fe defines max char and varchar length,see ScalarType;this is a temporary define,later I will use columns real length

kangpinghuang replied:
yes, for char type, it is a bit complex

wangbo committed:
Merge branch 'master_test'

gaodayue commented on be/src/olap/field.h:
```diff
@@ -57,6 +57,7 @@ class Field {
 
     inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
     inline void set_to_min(char* buf) const { return _type_info->set_to_min(buf); }
+    inline char* get_type_value_with_arena(Arena* arena) const { return _type_info->get_type_value_with_arena(arena); }
```
It's not necessary to copy all methods of TypeInfo to Field, I'd prefer to remove `get_type_value_with_arena` and `direct_copy_content` from this class.

wangbo replied:
Field defines column's content info,TypeInfo defines all type column's behavior;
In column_zone_map_builder,caller just need to care column abstract behavior,but not to known its realization.Different type has different realization.
Such as,when calling get_type_value_with_arena,int/double type not need data length,while string type need data length,so int/double type and char type has different method.
I think using Field's to realize polymorphism is better

gaodayue commented on be/src/olap/rowset/segment_v2/column_zone_map.h:
```diff
@@ -73,7 +73,8 @@ class ColumnZoneMapBuilder {
     ZoneMap _zone_map;
-    char* _max_string_value;
+    char* _max_char_value;
+    char* _max_varchar_value;
```
`_max_char_value` and `_max_varchar_value` are not used, should be removed

wangbo replied:
👌

gaodayue reviewed (changes requested):
LGTM generally. One thing I'm not sure is the change to varchar's set_to_max since it's been used by segment v1's zonemap. Would it break backward compatibility? @kangpinghuang Could you verify it?

wangbo commented on be/src/olap/types.h:
```diff
@@ -64,6 +64,10 @@ class TypeInfo {
         _direct_copy(dest, src);
     }
 
+    inline char* get_type_value_with_arena(Arena* arena) const {
```
👌

wangbo commented on be/src/olap/types.h:
```diff
@@ -587,13 +604,19 @@ struct FieldTypeTraits<OLAP_FIELD_TYPE_VARCHAR> : public FieldTypeTraits<OLAP_FI
     }
+    static char* get_type_value_with_arena(Arena* arena) {
+        char* type_value = arena->Allocate(sizeof(Slice));
+        Slice* real_type_value = (Slice*)type_value;
```
👌

wangbo commented on be/src/olap/types.h:
```diff
@@ -562,12 +572,19 @@ struct FieldTypeTraits<OLAP_FIELD_TYPE_CHAR> : public BaseFieldtypeTraits<OLAP_F
     }
+    static char* get_type_value_with_arena(Arena* arena) {
+        char* type_value = arena->Allocate(sizeof(Slice));
+        Slice* real_type_value = (Slice*)type_value;
```
👌

wangbo commented on be/src/olap/types.h:
```diff
@@ -587,13 +604,19 @@ struct FieldTypeTraits<OLAP_FIELD_TYPE_VARCHAR> : public FieldTypeTraits<OLAP_FI
         auto slice = reinterpret_cast<Slice*>(buf);
-        slice->size = 1;
-        memset(slice->data, 0xFF, 1);
+        memset(slice->data, 0xFF, slice->size);
```
👌

wangbo committed:
1. use allocate from
2. interpret not cast
3.remove useless var
4. set_to_min/max consist with master

wangbo committed:
refresh  commit

kangpinghuang commented:
I have check set_to_max in olap, I think original code has bug and this pr's code has fix it.
I check set_to_max used in two place of v1:
1. in zone map building. But for char/varchar type, there are no zone map for them, so there is no influence.
2. in split_range for query, I think the new max is larger than the old one, will not cause problems.
LGTM

kangpinghuang reviewed (dismissed):
LGTM

gaodayue reviewed (dismissed):
LGTM

wangbo commented:
> I have check set_to_max in olap, I think original code has bug and this pr's code has fix it.
> I check set_to_max used in two place of v1:
> 
> 1. in zone map building. But for char/varchar type, there are no zone map for them, so there is no influence.
> 2. in split_range for query, I think the new max is larger than the old one, will not cause problems.
>    LGTM

wangbo commented:
> LGTM

current commit is old version:in varchar's set_to_max,slice size set to 1,because set to max_string_value will cause UT core dump,I will see this case today

wangbo commented:
current commit is old version:in varchar's set_to_max,slice size set to 1
because set to max_string_value will cause delete_handler_test core dump,I will see this case today

imay commented on be/src/olap/rowset/segment_v2/column_reader.cpp:
```diff
@@ -426,6 +431,20 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
     RETURN_IF_ERROR(_reader->encoding_info()->create_page_decoder(data, options, &page->data_decoder));
     RETURN_IF_ERROR(page->data_decoder->init());
 
+    if (_reader->encoding_info()->encoding() == DICT_ENCODING) {
```
Better to add comment to explain why we create dict decoder lazily.

wangbo replied:
👌

imay commented on be/src/olap/types.h:
```diff
@@ -568,6 +577,13 @@ struct FieldTypeTraits<OLAP_FIELD_TYPE_CHAR> : public BaseFieldtypeTraits<OLAP_F
         auto slice = reinterpret_cast<const Slice*>(data);
         return HashUtil::hash(slice->data, slice->size, seed);
     }
+    static char* allocate_value_from_arena(Arena* arena) {
```
Now, Field has already contained the length of column.
I think it's more suitable to put this function in Filed other than here

wangbo replied:
agree;Field.h can decide allocate type value whether with size or not, but now ColumnZoneMapBuilder's field doesn't contains whole column info;
I think we can init a Field in SegmentWriter.init ,and then transfer the field to ColumnWriter and ColumnZonemap to instead of TypeInfo, but this need to change ColumnWriter constructor and ColumnZonemap, and change corresponding UT
So I prepare to submit another pr to fix the problem that string's ColumnZoneMap's max/min value init with real length.
Currently we first use fixed length.

imay replied:
OK, please don't forget delete this function then.

wangbo replied:
👌

wangbo commented:
> current commit is old version:in varchar's set_to_max,slice size set to 1
> because set to max_string_value will cause delete_handler_test core dump,I will see this case today

Core dump stack
>#0  SLL_Next (t=0xffffffffffffffff) at src/linked_list.h:45
>#1  SLL_TryPop (rv=<synthetic pointer>, list=0x436dce0) at src/linked_list.h:69
>#2  TryPop (rv=<synthetic pointer>, this=0x436dce0) at src/thread_cache.h:220
>#3  Allocate (oom_handler=0x25e5700 <tcmalloc::cpp_throw_oom(unsigned long)>, cl=5, size=64, this=<optimized out>) at src/thread_cache.h:379
>#4  malloc_fast_path<tcmalloc::cpp_throw_oom> (size=64) at src/tcmalloc.cc:1855
>#5  tc_new (size=64) at src/tcmalloc.cc:1976
>#6  0x0000000000c21672 in doris::WrapperField::WrapperField (this=0x4e07d40, rep=<optimized out>, variable_len=<optimized out>, 
>    is_string_type=<optimized out>) at /home/wangbo36/incubator-doris/be/src/olap/wrapper_field.cpp:84
>#7  0x0000000000c21b9e in doris::WrapperField::create (column=..., len=<optimized out>, len@entry=0)
>    at /home/wangbo36/incubator-doris/be/src/olap/wrapper_field.cpp:51
>#8  0x0000000000cb88e1 in doris::ColumnDataWriter::init (this=0x4bc4780) at /home/wangbo36/incubator-doris/be/src/olap/rowset/column_data_writer.cpp:72
>#9  0x0000000000ca273c in doris::AlphaRowsetWriter::_init (this=this@entry=0x4bf83c0)
>    at /home/wangbo36/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:275
>#10 0x0000000000ca3694 in doris::AlphaRowsetWriter::init (this=0x4bf83c0, rowset_writer_context=...)
>    at /home/wangbo36/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:73
>#11 0x0000000000c91b03 in doris::RowsetFactory::create_rowset_writer (context=..., output=output@entry=0x7fffcdf75318)
>    at /home/wangbo36/incubator-doris/be/src/olap/rowset/rowset_factory.cpp:50
>#12 0x0000000000bc9b3f in doris::TabletManager::_create_inital_rowset (this=this@entry=0x4d5a180, tablet=..., request=...)
>    at /home/wangbo36/incubator-doris/be/src/olap/tablet_manager.cpp:1250
>#13 0x0000000000bcc919 in doris::TabletManager::_internal_create_tablet (this=this@entry=0x4d5a180, alter_type=alter_type@entry=doris::SCHEMA_CHANGE, 
>    request=..., is_schema_change_tablet=is_schema_change_tablet@entry=false, ref_tablet=..., data_dirs=...)
>    at /home/wangbo36/incubator-doris/be/src/olap/tablet_manager.cpp:371
>#14 0x0000000000bcdd72 in doris::TabletManager::create_tablet (this=this@entry=0x4d5a180, request=..., stores=...)
>    at /home/wangbo36/incubator-doris/be/src/olap/tablet_manager.cpp:318
>#15 0x0000000000b82642 in doris::StorageEngine::create_tablet (this=<optimized out>, request=...)
>    at /home/wangbo36/incubator-doris/be/src/olap/storage_engine.cpp:760
>#16 0x0000000000b4fac3 in doris::TestDeleteConditionHandler::SetUp (this=0x4e08460)
>    at /home/wangbo36/incubator-doris/be/test/olap/delete_handler_test.cpp:173
>#17 0x0000000002530f37 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#18 0x000000000252ba9c in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#19 0x0000000002510df7 in testing::Test::Run() ()
>#20 0x00000000025116ce in testing::TestInfo::Run() ()
>#21 0x0000000002511d25 in testing::TestCase::Run() ()
>#22 0x0000000002518899 in testing::internal::UnitTestImpl::RunAllTests() ()
>#23 0x0000000002531eb3 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#24 0x000000000252c7a2 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#25 0x0000000002517570 in testing::UnitTest::Run() ()
>#26 0x00000000009fb316 in RUN_ALL_TESTS () at /home/wangbo36/incubator-doris/thirdparty/installed/include/gtest/gtest.h:2233
>#27 main (argc=<optimized out>, argv=0x7fffcdf75f88) at /home/wangbo36/incubator-doris/be/test/olap/delete_handler_test.cpp:951

wangbo committed:
add commit for lazy init dict

wangbo committed:
Merge remote-tracking branch 'remotes/upstream/master' into master_test

wangbo commented:
@imay please review again
1. add comment for dict lazy init
2.merge master branch to resolve conflict

imay reviewed (approved):
LGTM
